---
title: "Aurora Dataset"
output:
  html_notebook: default
  pdf_document: default
  html_document:
    df_print: paged
---

```{r libs,eval=TRUE, include=FALSE}
library(tidyverse)
library(DESeq2)
library(purrr)
library(ggplot2)
library(dplyr)
library(RColorBrewer)
library(pheatmap)
library(SummarizedExperiment)
library(knitr)
library(reshape2)
library(limma)
library(edgeR)
library(PCAtools)
library(tximeta)
library(org.Hs.eg.db)
library(biomaRt)
library(AnnotationDbi)
```

```{r}
#database and row data
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
attributes_first <- c("ensembl_gene_id_version", "external_gene_name", "description")
rowData <- getBM(attributes = attributes_first, 
                filters = "ensembl_gene_id_version",
                values = tx2gene$GENEID, 
                mart = ensembl)

#column data
colData_ak <- readxl::read_xlsx("~/Desktop/prj34865_metadata.xlsx")
one_metprim_DSBC <- readxl::read_xlsx("~/Desktop/phs002622.v1.pht012425.v1.p1.c3.AURORA_US_Metastatic_BC_Sample_Attributes.DS-BRCA.xlsx")
two_metprim_GR <- readxl::read_xlsx("~/Desktop/phs002622.v1.pht012425.v1.p1.c1.AURORA_US_Metastatic_BC_Sample_Attributes.GRU.xlsx")
three_metprim_HMB <- readxl::read_xlsx("~/Desktop/phs002622.v1.pht012425.v1.p1.c2.AURORA_US_Metastatic_BC_Sample_Attributes.HMB.xlsx")

DSBCGRHMB <- rbind(one_metprim_DSBC,two_metprim_GR,three_metprim_HMB)
joined_colData <- left_join(colData_ak, DSBCGRHMB, by = c("Sample Name"="SAMPLE_ID"))
joined_colData <- as.data.frame(joined_colData)

#get factor for the primary or met tumor
joined_colData$PRIMARY_METASTATIC_TUMOR <- ifelse(joined_colData$PRIMARY_METASTATIC_TUMOR == 'Not Applicable', 'NotApplicable',joined_colData$PRIMARY_METASTATIC_TUMOR)
joined_colData$PRIMARY_METASTATIC_TUMOR <- as.factor(joined_colData$PRIMARY_METASTATIC_TUMOR)


#database for gene info
txdb <- makeTxDbFromGFF("~/Downloads/gencode.v42.basic.annotation.gtf", 
                format = "auto", dataSource = NA, organism = "Homo sapiens", taxonomyId = NA)
k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")

#how i get gene symbol in it 
rowData <- rowData %>%
  dplyr::select(ensembl_gene_id_version, external_gene_name) %>%
  distinct(ensembl_gene_id_version, .keep_all = TRUE)
tx2 <- left_join(tx2gene, rowData, by = c("GENEID" = "ensembl_gene_id_version"))

#getting abundance files 
files <- file.path(dir, paste(sample_names, "_kallisto-quant/abundance.tsv", sep = ""))
names(files) <- paste0(sample_names)


#getting tximport file
cd1 <- tximport::tximport(files, type = "kallisto", txOut = TRUE)

cd$counts[1:6,]

rowdata <- tx2
tx2.1 <- rowdata[,1:2]
rownames(joined_colData) = joined_colData[["Run"]]

extra = setdiff(rownames(cd1[[1]]), as.character(tx2[["TXNAME"]]))

if (length(extra) > 0) {
    rowdata = rbind(rowdata, data.frame(TXNAME=extra, GENEID=extra, external_gene_name=extra))
}

rowdata = rowdata[match(rownames(cd1[[1]]), as.character(rowdata[["TXNAME"]])),]
rownames(rowdata) = rowdata[["TXNAME"]]
se = SummarizedExperiment(assays = list(counts = cd1[["counts"]], 
                                        abundance = cd1[["abundance"]], length = cd1[["length"]]),
                          colData = DataFrame(joined_colData),
                          rowData = rowdata)

#getting gene level counts 
gi = summarizeToGene(cd1, tx2gene = tx2.1)
gi$counts[1:6,1:3]

#get unique gene names and ids
growdata = unique(rowdata[,2:3])
growdata2 = growdata[match(rownames(gi[[1]]), growdata[["GENEID"]]),]
rownames(growdata) = growdata[["TXNAME"]]

#summarized experiment for count/
gse = SummarizedExperiment(assays = list(counts = gi[["counts"]], abundance = gi[["abundance"]], length = gi[["length"]]),colData = DataFrame(joined_colData),rowData = growdata)

#saving gse file
saveRDS(gse, file = "/Users/U1006746/Desktop/counts.AK.output/gene_counts.rds")

```


```{r data, eval=TRUE, include=FALSE}
gse <- readRDS(file = "/Users/U1006746/Desktop/counts.AK.output/gene_counts.rds")
list <- c()

```

```{r filter, eval=TRUE, include=FALSE}
#filter out genes (must have more than 2 counts) >= 3???
keep <- rowSums(cpm(gse) > 2) >= 3
se <- gse[keep,]
```

```{r setup, eval=TRUE, include=FALSE}
gene_info <- as.data.frame(rowData(se)@listData)
assay_info <- as.data.frame(assay(se))
sample_info <- as.data.frame(colData(se)@listData)

#how many pairs of tumors/metatstic we have
met_prim_pairs <- sample_info%>%
  select(Run,  submitted_subject_id,PRIMARY_METASTATIC_TUMOR, histological_type, body_site) %>%
  filter(PRIMARY_METASTATIC_TUMOR != "NotApplicable") %>%
  group_by(submitted_subject_id) %>%
  filter(any(PRIMARY_METASTATIC_TUMOR == "Primary") & any(PRIMARY_METASTATIC_TUMOR == "Metastatic"))
list <- met_prim_pairs$Run

#filter out columns (pairs)
se<- se[ , list] 
se$PRIMARY_METASTATIC_TUMOR <- factor(se$PRIMARY_METASTATIC_TUMOR, levels = c("Metastatic","Primary"))
se$histological_type <- factor(se$histological_type, levels = c("Not Available","Infiltrating Ductal Carcinoma","Ductal Carcinoma In Situ","Mixture of Tumor and Ductal Carcinoma In Situ","Infiltrating Lobular Carcinoma","Mixed Ductal and Lobular Carcinoma","Invasive Metaplastic Carcinoma with Squamous and Spindle Cell Differentiation", "Invasive Micropapillary Carcinoma","Poorly Differentiated Invasive Ductal Carcinoma" ))

treatment <- colData(se)$PRIMARY_METASTATIC_TUMOR
patient <- colData(se)$submitted_subject_id
body_site <- colData(se)$body_site
run <- colData(se)$Run
table(patient, treatment, body_site, run)

gene_info <- as.data.frame(rowData(se)@listData)
assay_info <- as.data.frame(assay(se))
sample_info <- as.data.frame(colData(se)@listData)
sample_info <- sample_info %>%
  arrange(Run)

# Replace NA values with "."
gene_info$external_gene_name <- ifelse(is.na(gene_info$external_gene_name), gene_info$GENEID, gene_info$external_gene_name)
gene_info$external_gene_name <- ifelse(gene_info$external_gene_name == "", gene_info$GENEID, gene_info$external_gene_name)
gene_info$external_gene_name <- ifelse(gene_info$external_gene_name == "Metazoa_SRP", "Metazoa_SRP.1", gene_info$external_gene_name)

#dds object
assays(se)$counts <- round(assays(se)$counts)
dds <- DESeqDataSet(se, design = ~ PRIMARY_METASTATIC_TUMOR)
dds <- dds[, list]
dds$PRIMARY_METASTATIC_TUMOR <- factor(dds$PRIMARY_METASTATIC_TUMOR, levels = c("Metastatic","Primary"))
```

```{r histology, eval=TRUE, include=FALSE}
#matching histology to proper subject 

hist <- data.frame(Subject_ID = met_prim_pairs$submitted_subject_id, Hist = met_prim_pairs$histological_type)
hist <- hist %>%
  filter(Hist != "Not Available")
body<- data.frame(Subject_ID = met_prim_pairs$submitted_subject_id, S = met_prim_pairs$body_site)

histology_body_site_data <- left_join(body, hist, by = "Subject_ID") 

sample_info <- sample_info %>%
  arrange(submitted_subject_id)

#creating new column data
hist2 <- histology_body_site_data %>%
  select(Subject_ID, Hist) %>%
  arrange(Subject_ID) %>%
  select(Hist)
sample_info2 <- cbind(sample_info, hist2)

```

For visualization purposes we use transformed counts.
```{r transform, eval=TRUE, include=FALSE}
#transforming 
#log transformed
se3 <- se
assay(se3) <- log2(assay(se3) + 1)
assay_info2 <- log2(assay_info +1)
```



```{r tables, eval=TRUE, include=FALSE}
#range from 0 - millions 
#most medians 100 - 300
dim(assay(se))
tibble(met_prim_pairs %>%
  group_by(submitted_subject_id) %>%
  count())
head(summary(assay(se)[,1:4]))
```

**Summary Statistics:**

* after removing low expressed genes and selecting only pairs of meta/prim tumors we are left with 29147 genes and 102 samples.
* corresponding to 39 subjects for those 102 samples
* most subjects have 2 samples, one subject has 9 samples (many metastitic sites)
* gene counts all range 0 - millions median around 200-300 (large variance)


```{r histograms, echo=FALSE}
#run/library size distribution scaled down
#distribution sum of samples scaled 
hist(colSums(assays(se)$counts/1e06), breaks = 10, xlab = "Sample Sum/1e06", main ="Distribution of Total Counts Per Sample")

hist(assays(se)$counts, breaks = 10,xlab = "Counts", main ="Distribution of Counts")
hist(assays(se)$counts,
     breaks=20000, xlim=c(0,10000), main="Distribution of Counts (truncated)",
     xlab="Counts (truncated axis)", 
     las=1, cex.axis=0.7)
hist(log2(assays(se)$counts +1), col="lightblue", border="white",
     breaks=100, main="Distribution of Counts (log)",
     xlab="Counts (log)",  
     las=1, cex.axis=0.7)

#closer look at one sample that has an outlier to see how data may be distributed
ggplot(assay_info) +
  geom_histogram(aes(x = SRR21068683), stat = "bin", bins = 100) +
  xlab("Counts") +
  ylab("Frequency") +
    ggtitle("Distribution of Gene counts for 1 sample") +
  xlim(0,100)

```
**Key take aways from counts per gene histograms:**

* most samples have around 30 million or 80 million total counts
* second histogram not helpful due to very few high counts
* middle shows range of raw counts truncated
* log transformed helps readability of counts.
* skewed distribution of gene expression counts for 1 sample examples/ not normal and count data (Poisson)


```{r boxplots, echo=FALSE}
#how treatments are distributed
par(mfrow = c(1,2))
boxplot(colSums(assays(se)$counts)/1e6 ~ treatment,ylab = "Scaled Total Gene Expression")
boxplot(colSums(assays(se3)$counts)/1e6 ~ treatment, ylab = "Scaled Total Gene Expression (log transformed)")
```

```{r subject, echo=FALSE}
#gene counts by subject
par(mfrow = c(1,2))
boxplot(colSums(assays(se)$counts)/1e6 ~ patient,las = 2, xlab = "",cex.axis = 0.8, ylab = "Scaled Total Gene Expression")
boxplot(colSums(assays(se3)$counts)/1e6 ~ patient, las = 2, xlab = "",cex.axis = 0.8, ylab = "Scaled Total Gene Expression (log transformed)")
```

```{r sample ,echo=FALSE}
#gene counts per sample
boxplot(log2(assay_info + 1), col=treatment, pch=".", 
        horizontal=TRUE, cex.axis=0.5,
        las=1, xlab="log2(Counts +1)")
```


```{r body,echo=FALSE}
#gene counts across different body sites 
par(mfrow = c(1,2))
boxplot(colSums(assays(se)$counts)/1e6 ~ body_site, las = 2, xlab = "",cex.axis = 0.7, ylab = "Total Gene Expression Per Sample")
boxplot(colSums(assays(se3)$counts)/1e6 ~ body_site, las = 2, xlab = "",cex.axis = 0.7, ylab = "Total Gene Expression Per Sample (log tranformed)")
```

```{r filter metandprim, echo=FALSE}
#pairs of met/primary
met_prim_pairs %>%
  ggplot(aes(x = submitted_subject_id)) +
  geom_bar(aes(fill = PRIMARY_METASTATIC_TUMOR)) +
  ggtitle("Subjects with Paired Prim/Met Samples")+
  xlab("Subject ID") +
  ylab("Count")+
  theme(axis.text.x = element_text(angle = 90),
        legend.title = element_text(size = 5))
```



```{r histology type, echo=FALSE}
histology_body_site_data %>%
  filter(S != "breast") %>%
  ggplot(aes(x = Hist, y = Subject_ID, fill= S)) +
  geom_tile() +
  theme(axis.text.x = element_text(angle = 45, size = 5, hjust = 1),
        legend.key.size = unit(0.25, "cm"),
        legend.title = element_text(size = 5),
        legend.text = element_text(size = 8)) 

```

**Key take aways from boxplots/sample visualizations: **

* log transformation helps variation
* Subject AER6 has most samples of metastatic tumors
* higher number met vs prim (to be expected)
* samples all have similar medians and ranges/variance
* brain has highest gene counts
* breast has largest range of gene counts 
* Infiltrating Lobular Carcinoma only one that metastasized to uterus
* Infiltrating Ductal Carcinoma most common histology type that met




```{r dispersion, echo=FALSE}
counts <- assays(se)$counts
variance.counts <- apply(assay(se),1,var)
mean.counts <-rowMeans(assay(se))
mean.var.col <- densCols(x=log2(mean.counts), y=log2(variance.counts))


plot(x=log2(mean.counts), y=log2(variance.counts),pch=16, cex=0.5, 
     col=mean.var.col, main="Log mean-variance relationship",
     xlab="Mean log2(normalized counts) per gene",
     ylab="Variance of log2(normalized counts)",
     panel.first = grid())
abline(a=0, b=1, col="brown")

```

**Key takeaways:**

* We just stabilized variance for visualization purposes about but not we will be looking at differential expression and will use our raw counts data. 
* Since this is "count data" we will be using a poisson distribution. 
* Our variance as seen in the figure above tends to increase with mean, has higher variance than mean most of the time.
* Above line so overdispersion (on log scale).
* We will therefore use Negative Binomial Distribution.



```{r dispersion2, echo=FALSE}
#using medain-of-ratio method to normalize
#estimate size factor for dds
dds = estimateSizeFactors(dds)

#estimating dispersion
dds = estimateDispersions(dds)

#plot dispersion
plotDispEsts(dds)
```

**Key takeaways:**

* this dispersion estimate plot shows the gene-wise estimates (variation of expression levels among different genes) (black), fitted values (red, MLE), and shrunken dispersion estimates (blue points). Circles are outliers.
* gene low read counts have higher dispersion
* not much difference between estimates and final estimates (not lot of variability between estimates)



**I. Primary vs Metastatic**
```{r diff, eval=TRUE, include=FALSE}
#DESeq DE
dds = nbinomWaldTest(dds)
contrast <- c("PRIMARY_METASTATIC_TUMOR", "Metastatic", "Primary")
resDESeq = results(dds, contrast = contrast)

resDESeq <- resDESeq[order(resDESeq$padj),]
head(resDESeq)
#summary of <0.05 p-value
summary(resDESeq, alpha = 0.05)

#just sig genes
padj.cutoff <- 0.05
sigRes <- resDESeq %>%
        data.frame() %>%
        filter(padj < padj.cutoff) %>%
        rownames_to_column(var = "gene")

#gene symbols in results table 
sigRes <- left_join(sigRes, gene_info, by = c("gene" = "GENEID"))
rownames(sigRes) <- sigRes$gene
sigRes <- sigRes[,-1]

#filter for heatmap
sigRes_fil <- sigRes[(sigRes$baseMean > 150) & abs(sigRes$log2FoldChange > 2),]

summary(sigRes_fil, alpha > 0.05)
```

```{r MA, echo=FALSE}
#blue points are statistically significant MA plots
DESeq2::plotMA(resDESeq, main = "DESeq2", ylim = c(-4, 4))
```

```{r HBB, echo=FALSE}
#which samples have highest counts of HBB gene (most differntially expressed, -4 log fold change)
gn.most.sign <- rownames(sigRes)[1]
gn.most.diff.val <- counts(dds, normalized=T)['ENSG00000244734.4',]
barplot(gn.most.diff.val, main='HBB', las=2, cex.names=0.5)
```

```{r p, echo=FALSE}
#seeing where p-values fall
hist(resDESeq$padj, breaks=20, col="grey", main="DESeq2 p-value distribution", xlab="DESeq2 P-value", ylab="Number of genes")
```
```{r heatmap, echo=FALSE}
#which genes to keep
genes.kept <- rownames(sigRes_fil)

#arrange by run
sample_info <- sample_info %>%
  arrange(Run)

#getting counts/treatmet/gene names situated
count.table <- counts(dds.norm, normalized=T)[genes.kept, ]
colnames(count.table) = paste(sample_info$PRIMARY_METASTATIC_TUMOR, sample_info$Run, sep = " : ")
count.table <- count.table %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene")
count.table <-  left_join(count.table, gene_info, by = c("gene" = "GENEID"))
rownames(count.table) <- count.table$external_gene_name
count.table <- count.table[,-103]
count.table <- count.table[,-1]

#heatmap z scores
count.table.z <- t(apply(count.table, 1, scale))
colnames(count.table.z) <- colnames(count.table)

#heatmap
h <- Heatmap(count.table.z, cluster_rows = T, cluster_columns = T, column_labels = colnames(count.table.z), name = "Z-score", row_labels = rownames(count.table.z))

```

```{r printing heatmap, echo=FALSE}
png("heatmap.png", res = 250, width = 5000, height = 5000)
print(h)
dev.off()
```

```{r PCA prim, echo=FALSE}
#PCA
vsd <- vst(dds, blind = TRUE)
p <- pca(assay(vsd), metadata = colData(vsd), removeVar = 0.1)
biplot(p,colby = "PRIMARY_METASTATIC_TUMOR", lab = NULL, legendPosition = 'right', legendTitleSize = 5, legendLabSize = 5)
```
```{r pca hist,echo=FALSE}
dds.hist<- DESeqDataSet(se, design = ~HISTOLOGICAL_TYPE)
vsd <- vst(dds.hist, blind = TRUE)
p <- pca(assay(vsd), metadata = colData(vsd), removeVar = 0.1)
biplot(p,colby = "HISTOLOGICAL_TYPE", lab = NULL, legendPosition = 'right', legendTitleSize = 5, legendLabSize = 4)
```



**Key Takeaway:**

* HBB, protein involved in making hemoglobin, is one of the most differential expressed between Met/Prim
* sample SRR21068811 and SRR21068806 have highest counts of HBB
* increased in Met tumors
* desirable p-value distribution
* LFC > 0 (up) there are 9024 genes, 31%
* LFC < 0 (down) there are 6346 genes, 22%
* in heatmap we see there are more decrease expression in primary more up in metastatic
* in PCA we see there is slight separation of primary vs metastatic however no real separation of different hematological/body types



**II. Histology Type**

```{r ductal,echo=FALSE, include=FALSE}
#getting histology type into the se col object
sample_info2 <- sample_info2 %>%
  arrange(Run)
se$HISTOLOGICAL_TYPE <- sample_info2$Hist
col_info <- as.data.frame(colData(se)@listData)
rownames(sample_info2) <- sample_info2$Run

#subset of se for just liver met
ductal<- c()
ductal_samples <- sample_info2 %>%
  filter(Hist %in% c("Infiltrating Ductal Carcinoma","Ductal Carcinoma In Situ"))
ductal <- ductal_samples$Run
se_ductal<- se[ ,ductal]

#dds object
dds.hist <- DESeqDataSet(se_ductal, design = ~HISTOLOGICAL_TYPE)
#estimating size factor
dds.hist = estimateSizeFactors(dds.hist)
#estimating dispersion
dds.hist = estimateDispersions(dds.hist)

```

```{r plots, echo=FALSE}
#plot dispersion
plotDispEsts(dds.hist)
```

```{r dds.hist,echo=FALSE, include=FALSE}
#deseq anaylsis
dds.hist = nbinomWaldTest(dds.hist)

#contrasts
contrasts_1 <- c("HISTOLOGICAL_TYPE", "Infiltrating Ductal Carcinoma","Ductal Carcinoma In Situ")
  
resHIST = results(dds.hist, contrast = contrasts_1)
```

```{r sigHist, echo=FALSE, include=FALSE}
#order by p-values
resHIST <- resHIST[order(resHIST$padj),]

#just sig genes
sigHist <- resHIST %>%
        data.frame() %>%
        filter(padj < padj.cutoff) %>%
        rownames_to_column(var = "gene")

#gene symbols in results table 
sigHist <- left_join(sigHist, gene_info, by = c("gene" = "GENEID"))
rownames(sigHist) <- sigHist$gene
sigHist <- sigHist[,-1]
```

```{r pvalue distribution2,echo=FALSE}
#seeing where p-values fall
hist(resHIST$padj, breaks=20, col="grey", main="P-value distribution", xlab="P-value", ylab="Number of genes")
```

```{r barplot, echo=FALSE}
#shows what samples have highest counts of the most differnetially expressed gene
gn.most.sign <- rownames(sigHist)[1]
gn.most.diff.val <-counts(dds.hist, normalized=T)["ENSG00000264940.5",]
barplot(gn.most.diff.val, main="SNORD3C", las=2, cex.names=0.5)

```

```{r trying heatmap,echo=FALSE, include=FALSE}
#heatmap
#filter for heatmap
sigHist <- sigHist[(sigHist$baseMean > 150) & abs(sigHist$log2FoldChange > 2),]

#which genes to keep
genes.kept <- rownames(sigHist)

#getting counts/treatmet/gene names situated
#get right hist names for columns we have
hist.count.table <- counts(dds.hist, normalized=T)[genes.kept, ]
cols <- colnames(hist.count.table)
sample_info_histnames <- ductal_samples %>%
  filter(rownames(ductal_samples) %in% cols)

colnames(hist.count.table) = paste(colnames(hist.count.table), sample_info_histnames$Hist , sep = " : ")
hist.count.table <- hist.count.table %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene")
hist.count.table <-  left_join(hist.count.table, gene_info, by = c("gene" = "GENEID"))
rownames(hist.count.table) <- hist.count.table$external_gene_name
hist.count.table <- hist.count.table[,-104]
hist.count.table <- hist.count.table[,-1]

#heatmap z scores
hist.count.table.z <- t(apply(hist.count.table, 1, scale))
colnames(hist.count.table.z) <- colnames(hist.count.table)

#heatmap
h2 <- Heatmap(hist.count.table.z, cluster_rows = T, cluster_columns = T, column_labels = colnames(hist.count.table.z), name = "Z-score", row_labels = rownames(hist.count.table.z))

```

```{r trying2,echo=FALSE}
png("heatmap2.png", res = 250, width = 5000, height = 7000)
print(h2)
dev.off()
```

**Key Takeaways:** 

* fewer "significant" p-values
* highest expression of SNORD3C in "Infiltrating Ductal Carcinoma"
* SNORD3C associated with invasive BC
* my sample sizes are very different and one is too small so this will not work


